CI: Build docker image if environment changes

ewels opened the pull request:
WIP for new GitHub actions to build the docker image in the CI if the environment changes.

Building the docker image is fairly easy, it was getting the logic in place to only do it when we need to that was a little more fiddly. I got that working with the [`technote-space/get-diff-action@v1`](https://github.com/marketplace/actions/get-diff-action) action in the end.

If either `environment.yaml` or `Dockerfile` have been edited, then a new docker container is built using `docker build . -t nfcore/rnaseq:dev`.

If `environment.yaml` or `Dockerfile` have _not_ been edited, then the workflow just pulls the image with `docker pull nfcore/rnaseq:dev`

## Examples

* `push` actions run where the environment was not changed (`docker pull`):
    * https://github.com/ewels/nf-core-rnaseq/runs/730898163?check_suite_focus=true
* `pull_request` actions run where the environment was changed (`docker build`):
    * https://github.com/ewels/nf-core-rnaseq/pull/32/checks?check_run_id=730914424

## Caveats

The workflow runs on `push` and `pull_request`. The latter compares the `git diff` over the range of changes, but `push` just gets that one commit. So if you change the environment in one commit and then change something else in the second then the second `push` event will fail as it will pull the image because the environment files didn't change in that second, isolated commit. You can see this by looking at the latest actions `push` run on the demo PR branch above [here](https://github.com/ewels/nf-core-rnaseq/pull/32/checks?check_run_id=730903853). Pull-requests should work fine.

One way to get around this is to _always_ build the docker image for every single CI run. However, this is quite a bit slower. Using the above actions runs as an example, building the image took around 7-8 minutes. Pulling the image took about 1 minute. The entire run-time varies a bit according to the test, but the docker build takes about half of the total run time, so it's quite slow.

Another option is to always push to DockerHub with a new tag name corresponding to the branch name. Then use this in the pull. I think that this should work, though it will then require (a) every fork to have a corresponding Docker Hub repo set up and (b) will pollute with a _lot_ of Docker Hub tags.

## To-Do

- [x] Push to Docker Hub
- [ ] Update linting so that it doesn't fail
- [ ] Update `nf-core bump-versions` to edit the `docker build -t` tag


---

## PR checklist
 - [x] PR is to `dev` rather than `master`
 - [x] This comment contains a description of changes (with reason)
 - [ ] If you've fixed a bug or added code that should be tested, add tests!
 - [ ] If necessary, also make a PR on the [nf-core/rnaseq branch on the nf-core/test-datasets repo]( https://github.com/nf-core/test-datasets/pull/new/nf-core/rnaseq)
 - [ ] Ensure the test suite passes (`nextflow run . -profile test,docker`).
 - [ ] Make sure your code lints (`nf-core lint .`).
 - [ ] Documentation in `docs` is updated
 - [x] `CHANGELOG.md` is updated
 - [ ] `README.md` is updated

ewels commented:
NB: 8 minutes is still ridiculously fast compared to the very many _hours_ that DockerHub usually takes to build..

apeltzer commented:
Honestly, I think the issue you mentioned in the caveat section is negligible. We only have to take care that whenever the environment/docker stuff is updated, that this only contains a single commit (or multiple that change environment/dockerfile). This should be achievable: In the worst case one could close the PR / squash commits together in such cases or not?

Great job - the filtering approach for git diff looks like a cool thing to do/use here!

apeltzer commented:
Another Q: What do we do with multiple containers, e.g. in the Sarek use case?

MaxUlysse commented:
> Another Q: What do we do with multiple containers, e.g. in the Sarek use case?

Building the container with CircleCI in Sarek case, because of huge memory available there.
But I guess we could add some specific checks towards a specific `environment.yml` / `Dockerfile` pair of files

drpatelh commented:
Wicked! Will it matter if either `environment.yml` / `Dockerfile` dont exist?

ewels commented:
Not quite.. As it stands, tests on most pull-requests will fail unless they are a single commit. A PR CI run typically has two sets of tests:

* The `pull_request` event tests
    * Takes the diff range and merges it to the upstream code - these  will pass
* The `push` event tests
     * The tests running on the unmerged code, from the last commit, I think - these will fail

Once merged, everything should pass (squash merge or regular merge commit). The resulting `push` event will be on a commit where the files are edited, so the docker image will build again. The only one we should avoid is a rebase merge.

ewels commented:
> Wicked! Will it matter if either environment.yml / Dockerfile dont exist?

No, it will just not build the new container. It filters the diff list for these filenames, so they'll just never show up as being edited.

ewels commented:
> Another Q: What do we do with multiple containers, e.g. in the Sarek use case?

Currently either do nothing, or break, I'm not sure. The filters I think will only match `environment.yaml` and `Dockerfile` in the root, but I haven't tested this. I'm I'm right then it will just ignore any others and do a docker pull as normal. If I'm wrong and other ones are picked up, it will probably break as it will try to do a `docker build` in the root directory.

ewels commented:
Note - for cases where we have multiple docker images (eg. nf-core/modules) I have already written another GitHub actions script that works for this. See https://github.com/ewels/nf-core-modules/blob/docker-github-packages/.github/workflows/docker.yml

It's a beauty. Could probably be simplified quite a bit / made prettier now that I have discovered the `technote-space/get-diff-action@v1` action...

MaxUlysse commented:
> > Another Q: What do we do with multiple containers, e.g. in the Sarek use case?
> 
> Building the container with CircleCI in Sarek case, because of huge memory available there.
> But I guess we could add some specific checks towards a specific `environment.yml` / `Dockerfile` pair of files

Which is actually how these containerss are handled at the moment in a specific test
https://github.com/nf-core/sarek/blob/75fe254f0b5d93d1f3f0173f2163afa83686f6fd/.github/workflows/ci.yml#L29-L51

MaxUlysse commented on .github/workflows/ci.yml:
```diff
@@ -5,26 +5,44 @@ on: [push, pull_request]
+        if: ${{ !env.GIT_DIFF }}
+        run: |
+          docker pull nfcore/rnaseq:dev
+          docker tag nfcore/rnaseq:dev nfcore/rnaseq:dev
```
Can we finally get rid of this pull / tag?

ewels replied:
No - currently this is used if the environment doesn't change, same behaviour as before. See description about it taking 1 minute instead of 8 minutes etc...

MaxUlysse replied:
I meant can we have just a pull instead of a pull / tag

If the environment doesn't change, then we're pulling dev.

Except for a release, where we're building a new container anyway

ewels replied:
Whether this will work or not depends on which approach we take. See https://github.com/nf-core/rnaseq/pull/423#issuecomment-637569636

* If we take approach (1) then we can drop the `docker pull` completely.
* If we take approach (2) then we have much more complex tagging and still need this line to make the pipeline code match the environment.

So, no we can't remove it as it stands I think 😄

apeltzer replied:
We might want to make a separate issue out of this @MaxUlysse and then proceed / try to find out if we can get rid of this then ...?

MaxUlysse commented:
I will just add a step to push to dockerhub if tests are successful.

ewels commented:
> I will just add a step to push to dockerhub if tests are successful.

Yeah I will do once we have decided how to proceed (see TODO checklist in original comment).

MaxUlysse commented:
> > I will just add a step to push to dockerhub if tests are successful.
> 
> Yeah I will do once we have decided how to proceed (see TODO checklist in original comment). I didn't want to do this right away in case we decide to go for option 3 of pushing to dockerhub for every branch / commit that changes the environment.

Currently DockerHub rebuilt the container for every PR merged.

I think if we go towards just pushing to DockerHub whenever we change `environment.yml`/`Dockerfile`, we will be making less building, so I'd say, let's go for it

MaxUlysse commented on .github/workflows/ci.yml:
```diff
@@ -5,26 +5,44 @@ on: [push, pull_request]
+
+      - name: Pull docker image
+        if: ${{ !env.GIT_DIFF }}
+        run: |
+          docker pull nfcore/rnaseq:dev
+          docker tag nfcore/rnaseq:dev nfcore/rnaseq:dev
```
```suggestion
        run: docker pull nfcore/rnaseq:dev
```

MaxUlysse replied:
This is what I meant ;-)

ewels replied:
Yup, I understand - see https://github.com/nf-core/rnaseq/pull/423#discussion_r433910287 for why it wouldn't work.

MaxUlysse commented on .github/workflows/ci.yml:
```diff
@@ -5,26 +5,44 @@ on: [push, pull_request]
+
+      - name: Build new docker image
+        if: env.GIT_DIFF
+        run: docker build . -t nfcore/rnaseq:dev
```
Just need to have a condition here for release (or have that in bump-version for the release)

ewels replied:
Yeah, I think that this will need to be changed in the `bump-version` code, as the code versions change in more CI events than just the release.

ewels replied:
Added to the TODO in the top comment 👍

ewels commented:
Ok, this thread has become a bit spread thin. The TL;DR; is this:

Implementation in this PR is a good start, but flawed because of the issue with detecting changes on only the single commit for the `push` event

I can think of two potential solutions, which we need to choose from:

1. Just drop the `docker pull` stuff and _always_ build the docker image for every single CI run
  * Up-side: Simple, should work
  * Down-side: Adds about 8-10 minutes extra time on to every CI run
2. If the environment changes then build and always push to DockerHub with tag based on the branch name. Pull from this specific tag if the environment does not change.
  * Up-side: Skips unnecessary builds, saves some time on most CI runs
  * Down-side: More complicated, means we end up with tonnes of Docker Hub tags
  * Bad down-side: Not sure how this will work with forks. Probably every fork user will need to set up a DockerHub repo for their fork 😞

ewels commented:
Updated PR to always build docker image for everything. Feels wasteful building the same image so many times, but it's the only way I can figure out to make everything work easily.

Added code to push container to Docker Hub if a `push` event on `dev` or a `release` event (tagged with release number and also `:latest`).
This bit is untested - mostly just copied from [Sarek](https://github.com/nf-core/sarek/blob/master/.circleci/config.yml) 😉

MaxUlysse commented:
I'm trying to do a conditional build:
https://github.com/nf-core/sarek/pull/213

apeltzer commented on .github/workflows/ci.yml:
```diff
@@ -1,30 +1,85 @@
+
+      - name: Build new docker image
+        if: env.GIT_DIFF
+        run: docker build . -t nfcore/rnaseq:dev
```
This is completely separated everytime I guess? So no need to add a `–no-cache` here ?

ewels replied:
Yes, should be completely separate every time. But then it doesn't hurt to add a `--no-cache` either, so may as well do so I guess. Added 👍

apeltzer commented on .github/workflows/ci.yml:
```diff
@@ -1,30 +1,85 @@
+        uses: actions/checkout@v2
+
+      - name: Build new docker image
+        run: docker build . -t nfcore/rnaseq:dev
```
Maybe add the `--no-cache` to be safe?

ewels replied:
Done 👍

ewels replied:
(Shouldn't be needed, as jobs are totally separate. But then doesn't hurt either)

ewels commented:
Updated to only run pipeline tests on `push` if on nf-core `dev` tag. This should only be when PRs are merged, so that we can push a new container to Docker Hub.

This means that we are no longer testing the pipeline run on every commit on branches / forks, but I'm not sure that the results of these tests are used _so_ much anyway, so maybe this isn't a huge loss.

Dropping the `push` event like this solves the problem described in https://github.com/nf-core/rnaseq/pull/423#issuecomment-637557770

ewels commented:
✅ Only linting runs on [my fork](https://github.com/ewels/nf-core-rnaseq/commits/actions-build-docker), no pipeline tests
✅ CI is pulling the existing docker image on this PR, as I didn't update the environment

Once this PR is merged I'll make a new one to update MultiQC, then we can test whether it builds and pushes the new container as expected.

ewels commented:
✅ Disabled [automated builds](https://hub.docker.com/repository/docker/nfcore/rnaseq/builds) on Docker Hub for the rnaseq pipeline

ewels commented:
🤩 
![image](https://user-images.githubusercontent.com/465550/83753365-6b96a380-a66a-11ea-86e7-29b5dae24300.png)
